editor/plugins/canvas_item_editor_plugin.cpp

```diff
@@ -2626,6 +2665,83 @@ void CanvasItemEditor::_draw_grid() {
 	}
 }
 
+void CanvasItemEditor::_draw_ruler_tool() {
+
+	if (tool != TOOL_RULER)
+		return;
+
+	bool is_snap_active = snap_active ^ Input::get_singleton()->is_key_pressed(KEY_CONTROL);
+
+	if (ruler_tool_active) {
+		Color ruler_primary_color = get_color("accent_color", "Editor");
+		Color ruler_secondary_color = ruler_primary_color;
+		ruler_secondary_color.a = 0.5;
+
+		Point2 begin = ruler_tool_origin - view_offset * zoom;
+		Point2 end = snap_point(viewport->get_local_mouse_position() / zoom + view_offset) * zoom - view_offset * zoom;
+		Point2 corner = Point2(begin.x, end.y);
+		Vector2 length_vector = (begin - end).abs() / zoom;
+
+		bool draw_secondary_lines = (begin.y != corner.y && end.x != corner.x);
+
+		viewport->draw_line(begin, end, ruler_primary_color, 3);
```

Comment:
Line widths should be multiplied by `EDSCALE` and rounded:

```cpp
Math::round(EDSCALE * 3)
```

Also, you can add a `true` parameter at the end to enable antialiasing. I recommend doing this for all lines that aren't drawn at straight angles (i.e. that may be drawn diagonally at some point).

The same applies to the two other `draw_line()` calls below.